src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs

```diff
@@ -97,16 +97,16 @@ private void AppendContentLength(ReadOnlySpan<byte> value)
 
         [MethodImpl(MethodImplOptions.NoInlining)]
         [SkipLocalsInit]
-        private void AppendContentLengthCustomEncoding(ReadOnlySpan<byte> value, Encoding? customEncoding)
+        private void AppendContentLengthCustomEncoding(ReadOnlySpan<byte> value, Encoding customEncoding)
         {
             if (_contentLength.HasValue)
             {
                 KestrelBadHttpRequestException.Throw(RequestRejectionReason.MultipleContentLengths);
             }
 
             // long.MaxValue = 9223372036854775807 (19 chars)
-            Span<char> decodedChars = stackalloc char[20];
-            var numChars = customEncoding!.GetChars(value, decodedChars);
+            Span<char> decodedChars = stackalloc char[32]; // power of two
```

Comment:
Curious, why is using a power of two better?

Reply:
I don't know the exact rationale of the JIT here, but by [looking at the codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9JqwAqMBBg3VNAZiakGAMQgQGAbxoMnTM2AAW2KAwCCACgCUjs4O1M5hDLJi2FwAPO6eAHwMuFFcDAC8yRjYYADW2JKSEGAM8VAA2qR0ALpG4c7EAOzJqeU1dc4AvkFOPS6lHl4AQgF9IfVOkdFxg0kp0RlZOfmFxQOe5SaktX1hTS3RbTuhXX195bK5vGIAMsUFuACSXLwY1X3EroM+5KMnTuMJlNYmU5qlFrhsnkCkUSmVKu1dg1mvMuEcOk5uv8GOdLtc7mAHs9Xu9sZ91sNfoFsYD6sCZokDmlMpDljC1vCtscJkwUa1EdisZ0gA===) a power of two is better.

Here I believe an optimization comes into play, that treats small stack alloc sizes as normal locals.
So maybe that's also the reason why SkipLocalsInit has no effect when 20 is used size. With a power of two, the zeroing isn't done (as intentend). In the sharplab linked above you can try to increase the allocation size (e.g. to 120 and 128) and see that there's a difference between small sizes and larger ones.

Reply:
Interesting. I don't think that `// power of two` explains this sufficiently though. Playing with the SharpLab, it looks like the extra, non-SkipLocalsInit affected stackalloced spans happen betweens 33-48 bytes which 20 chars happens to fall in-between. @gfoidl Do you know if there's a runtime issue tracking this?

I think we should only ask for what we need and let the JIT optimize how it sees fit.

Reply:
> 33-48 bytes

The sharplab is for `char`, so 66-96 bytes. I can't find any relevant issue in dotnet/runtime, but I remember there were some changes in the way how stack space is zeroed with various threshold and 96 bytes being one of them (IIRC).

For a proper reasoning what's going on here we should ask JIT-team. Maybe it's a bug, or it's by intention.
At least it's a bit surprising, especially that SkipLocalsInit hasn't the effect when not a power of two and small allocation sizes.

> we should only ask for what we need and let the JIT optimize how it sees fit.

The span is sliced a bit later to the correct size anyway, so the size is in excess at time of allocation. But your view is more "natural" to rounding up to a power of two.
BTW: In dotnet/runtime almost always a power of two is used for stackalloc.

Reply:
I'll revert this for now since it's unrelated to this PR. I recommend following up with the JIT folks to clarify the recommendations here.

Reply:
> The sharplab is for char, so 66-96 bytes. 

I was testing with `stackalloc byte[N]`.

Reply:
Filed https://github.com/dotnet/runtime/issues/55001 to track this.